### PR TITLE
[FIX] introduce "title" as mandatory MD for series

### PR DESCRIPTION
**Pull request by chfsx**
This PR should fix the issue in #328 

Currently it's possible to delete or never configure the metadata field "title" for series. This leads to errors on both sides, ILIAS and even opencast.

The fix introduces an update step which creates the missing field if it doesn't exist. beside that, the MD definition now holds a "mandatory" flag und the delete action is no longer available for mandatory fields.

### CODE REVIEW

**Comment by dagraf** [comment]
Thx for the PR. I appreciate the fix and can confirm the changes.

Therefore and @ferishili : Can you please review this PR and merge it if everything is fine or comment it, if not? Thx!

**Comment by Amstutz** [comment]
Hey @chfsx 

Thx a lot for the changes. I can only superficially check the changes, however to me they look good. Maybe @ferishili can have an other glance. 

Special thx for improving the overall code quality with the PR with some minor neat little refactorings (such as types).

File: classes/Conf/Metadata/class.xoctMetadataConfigGUI.php [hunk 1]
```diff
@@ -222,7 +226,7 @@ protected function loadList(): void
     /**
      * Converts or better say digest the loaded list and replaces the translation to be processed by the plugin.
      *
-     * @param array  $raw_list the raw list that comes from the listprovider endpoints
+     * @param array $raw_list  the raw list that comes from the listprovider endpoints
```
Comment (ferishili) [inline classes/Conf/Metadata/class.xoctMetadataConfigGUI.php]: Would be nice to remove extra space!

```suggestion
     * @param array $raw_list the raw list that comes from the listprovider endpoints
```

File: classes/Setup/class.ilOpenCastDBUpdateSteps.php [hunk 2]
```diff
@@ -55,4 +55,31 @@ public function step_2(): void
...
+
+        $next_id = $this->db->nextId('xoct_md_field_series');
+
+        $this->db->insert(
+            'xoct_md_field_series',
+            [
+                'id' => ['integer', $next_id],
+                'field_id' => ['text', 'title'],
+                'title_de' => ['text', 'Titel'],
+                'title_en' => ['text', 'Ttitle'],
```
Comment (ferishili) [inline classes/Setup/class.ilOpenCastDBUpdateSteps.php]: "Ttitle" or Title?
```suggestion
                'title_en' => ['text', 'Title'],
```

File: src/Model/Metadata/Definition/MDCatalogueFactory.php [hunk 3]
```diff
@@ -38,7 +38,7 @@ public function event(): MDCatalogue
     public function series(): MDCatalogue
     {
         return $this->series_catalogue ?? $this->series_catalogue = new MDCatalogue([
-            new MDFieldDefinition(MDFieldDefinition::F_TITLE, MDDataType::text(), false, true),
+            new MDFieldDefinition(MDFieldDefinition::F_TITLE, MDDataType::text(), false, true, true), // Title is Mandatory for Series
```
Comment (ferishili) [inline src/Model/Metadata/Definition/MDCatalogueFactory.php:41]: I like the Mandatory option here.
That would be good to have that also for Event in terms of consistency in
 https://github.com/opencast-ilias/OpenCast/blob/a0392d8852b42c777aed9eb6eba966c2fd38cb5c/src/Model/Metadata/Definition/MDCatalogueFactory.php#L19

**Review by ferishili — CHANGES_REQUESTED** [review]

**Comment by ferishili** [comment]
Hi @chfsx,
Thanks for your work and also the improvements. Everything works fine.
I added a few minor changes.

File: src/Model/Metadata/Definition/MDCatalogueFactory.php (hunk 3 above)
Comment (chfsx) [reply src/Model/Metadata/Definition/MDCatalogueFactory.php:41]: Thanks, I was thinking about that as well. But I'm not sure if its really mandatory for events as well (on the Opencast-side), does anybody know?

**Comment by chfsx** [comment]
Thanks for the findings and suggestions! I merged them where possible, concerning the mandatory title in event: I just lack of opencast knowledge there...

File: src/Model/Metadata/Definition/MDCatalogueFactory.php (hunk 3 above)
Comment (ferishili) [reply src/Model/Metadata/Definition/MDCatalogueFactory.php:41]: Of course, it is mandatory for events too, however, the External API will return an error if you forget it, unlike Series!
So I would say, it is safe to put that there as well

Comment (ferishili) [reply src/Model/Metadata/Definition/MDCatalogueFactory.php:41]: Ping! @chfsx

**Comment by chfsx** [comment]
@ferishili thanks for the feedback concerning events, I added another commit to make this mandatory as well

**Comment by ferishili** [comment]
> @ferishili thanks for the feedback concerning events, I added another commit to make this mandatory as well

Hi @chfsx,
I cannot see your commit/any changes regarding this!

File: classes/Setup/class.ilOpenCastDBUpdateSteps.php [hunk 4]
```diff
@@ -82,4 +82,31 @@ public function step_3(): void
             ]
         );
     }
+
+    public function step_4(): void
+    {
+        // check for missing mandatory "title" MD for series
```
Comment (ferishili) [inline classes/Setup/class.ilOpenCastDBUpdateSteps.php]: This step would not be necessary, but as for a double-checker I would say, that is ok.
A small typo:
```suggestion
        // check for missing mandatory "title" MD for event
```

**Review by ferishili — COMMENTED** [review]
Thanks for the changes, it looks good to me. As soon as the suggestion is applied, this could be merged.

File: classes/Setup/class.ilOpenCastDBUpdateSteps.php (hunk 4 above)
Comment (ferishili) [reply classes/Setup/class.ilOpenCastDBUpdateSteps.php]: @chfsx Ping! 
The small suggestion could be applied, and then we could get it merged!

**Comment by chfsx** [comment]
Thanks a lot, I will rebase, merge and squash locally and release e new version which includes other fixes as well

**Review by ferishili — APPROVED** [review]

**Comment by chfsx** [comment]
This has been merged locally and is part of v8.2.8

**Pull request closed without merging**